_posts/2020-07-21-Securing-Applications-Azure-AD.md

```diff
@@ -0,0 +1,405 @@
+</div>
+
+This is where we need to tweak some configuration settings for the library to work with Azure AD.
+`strictDiscoveryDocumentValidation` needs to be disabled due to the fact that not all URLs in the discovery document start with the issuer URL. 
+This makes strict parsing fail, so we disable it.
+
+You might also have noticed the weird looking `api://<<<client_id>>>/app` value in the list of scopes. 
+The reason why we do this is explained very well in this [Medium blogpost](https://medium.com/@abhinavsonkar/making-azure-ad-oidc-compliant-5734b70c43ff){:target="_blank" rel="noopener noreferrer"} but boils down to the fact Azure AD uses a nonce in a special way in its JWT header.
+This breaks the standard JWT validation.
+If we include an application specific scope here, this will no longer be the case. 
+Our Angular application won't actually care for this as it just passes access tokens to the Spring Boot back-end.
+The validation there will fail, resulting in a 401: Unauthorized. 
+You can define this scope in the Azure Portal, under `Expose an API` > `Add a scope`.
+
+<div style="text-align: center;" >
+  <img class="image fit-contain" src="{{ '/img/azure-ad/add-scope.png' | prepend: site.baseurl }}" alt="" width="60%" />
+</div>
+
+> This application-specific scope can have any name, so it doesn't have to be `app`. 
+> Just make sure you use the same scope in the application as the one you defined in the Azure Portal.
+
+Another requested scope we configure is `openid`.
+This indicates that we also want to log in the user.
+We will not only receive an access token to contact the back-end API, but also an id token with information about the logged-in user.
+
+> Azure AD will serve an id token, regardless of the open-id scope. But we include it anyway to respect the [specification](https://openid.net/specs/openid-connect-core-1_0.html){:target="_blank" rel="noopener noreferrer"}.
+
+The next step is to trigger the login when a user has not logged in yet:
+
+```typescript
+// app.component.ts
+constructor(private oauthService: OAuthService) {
+    this.oauthService.configure(authCodeFlowConfig); // (1)
+    this.oauthService.loadDiscoveryDocumentAndLogin(); // (2)
+
+    this.oauthService.set-upAutomaticSilentRefresh(); // (3)
+}
+```
+
+1. We set up the OAuthService with the configuration from the previous step.
+This makes sure it uses the authorization code flow + PKCE with the correct parameters.  
+2. The discovery document will be loaded, which is the issuer URI plus the `.well-known/openid-configuration` suffix and then start the login process.  
+3. As access tokens have a short lifespan, we want them to be automatically refreshed in the background.
+
+We can now try out the application and should be redirected to the Microsoft login page.  
+After logging in, when we browse to the heroes page, we can see the 401 is gone, and the heroes are fetched again. 
```

Comment:
Maybe also mention and link the "final" branch of both the backend and the frontend to share the full solution and to conclude the blogpost.